### ISSUE TITLE
All page object fields are null

### ISSUE THREAD

**Issue by chalin** [dart-lang/angular#783]
The component tests of our [AngularDart example apps](https://github.com/dart-lang/site-webdev/tree/master/examples/ng/doc) are failing because **[pageloader][] page object fields are all null** (they aren't being bound).

The same tests **were running fine under Dart 2.0-dev.16 and Angular 5.0.0-alpha+2.** They now fail under Dart 2.0-dev.17 and Angular 5.0.0-alpha+3 (MacOS 10.13.2). This is what an attempt at running [toh-1][] tests looks like:

```nocode
toh-1 > pub run angular_test --test-arg=--tags=aot --test-arg=--platform=chromeheadless --test-arg=--reporter=expanded --verbose --serve-arg=--port=8081 --experimental-serve-script=none
Using external server running on port 8081.
Running tests...
pub run test --tags=aot --platform=chromeheadless --reporter=expanded --pub-serve=8081
00:00 +0: compiling test/app_test.dart
"pub serve" is compiling test/app_test.dart...
00:00 +0: test/app_test.dart: title
00:00 +0 -1: test/app_test.dart: title [E]
NullError: Cannot read property 'visibleText' of null

00:00 +0 -1: test/app_test.dart: initial hero properties
00:00 +0 -2: test/app_test.dart: initial hero properties [E]
NullError: Cannot read property 'visibleText' of null

00:00 +0 -2: test/app_test.dart: update hero name
00:00 +0 -3: test/app_test.dart: update hero name [E]
NullError: Cannot read property 'type' of null

00:00 +0 -3: Some tests failed.
```

Notice the: **NullError: Cannot read property 'visibleText' of null**. I get the same result if I use `chrome` as platform. Here is the transcript from pub serve:

```nocode
toh-1 > pub serve web test --web-compiler=dartdevc
Loading source assets... 
Loading angular, test/pub_serve and dart_to_js_script_rewriter transformers... 
Serving angular_tour_of_heroes web  on http://localhost:8080
Serving angular_tour_of_heroes test on http://localhost:8081
Build completed successfully
[test] GET app_test.dart.browser_test.dart.js.map → angular_tour_of_heroes|test/app_test.dart.browser_test.dart.js.map
[test] GET Served 3 assets.
[test] GET Served 6 assets.
[test] GET dart_sdk.js → angular_tour_of_heroes|test/dart_sdk.js
[test] GET Served 66 assets.
```

As an aside, the tests only run from the command line (as shown above) after I remove the HTML test harnes file: `rm test/app_test.html`.

[pageloader]: https://pub.dartlang.org/packages/pageloader
[toh-1]: https://github.com/dart-lang/site-webdev/tree/master/examples/ng/doc/toh-1

cc @kwalrath @matanlurey @kevmoo

**Comment by matanlurey** [comment]
@kevmoo nothing has changed on our side, is this a .17 regression?

**Comment by kevmoo** [comment]
> nothing has changed on our side, is this a .17 regression?

If the only delta is going from .16 to .17, then I'd guess yes.

**Comment by chalin** [comment]
No, the working combo was dev.16 & Angular 5.0.0-alpha+2 vs. dev.17 & alpha+3.

**Comment by matanlurey** [comment]
Internally we are still on dev.13 but with ~alpha+3, so unfortunately I don't know.

That being said our travis tests are green, I think, and we have at least one PO test.

**Comment by chalin** [comment]
I'll test that combo tomorrow.

**Comment by chalin** [comment]
Ok, this took a while to track down.

**TL;DR** tests fail under **dartdevc**.

Here is what I tried, in order:

### 1. Travis: Last successful test run combo ✅

[This Travis log](https://travis-ci.org/dart-lang/site-webdev/jobs/327968236) is from the last time component **tests ran successfully on Travis**. Setup:

- SDK **2.0.0-dev.16**
- Angular 5.0.0-**alpha+2**
- Web-compiler: **dart2js**
- Test platform: **content-shell**

### 2. Local (MacOS): successful setup (after patching generated code) ✅

When I try this combo locally under MacOS, it fails, but I recalled that I've been unable to run 5.x tests locally for months now (I've been relying on Travis, and expected the problem to go away, which it does in 2.0-dev.17). 

This is the error that is reported when I try to run tests locally:

```nocode
pub run angular_test --test-arg=--tags=aot --test-arg=--platform=content-shell --test-arg=--reporter=expanded --verbose --serve-arg=--port=8081 --experimental-serve-script=none
Using external server running on port 8081.
Running tests...
pub run test --tags=aot --platform=content-shell --reporter=expanded --pub-serve=8081
00:00 +0: test/app_test.dart: title
00:00 +0 -1: test/app_test.dart: title [E]
'package:angular/src/core/linker/app_view.dart': error: line 604 pos 53: ',' or ')' expected
void Function(E) eventHandler0<E>(void Function() handler) {
^

package:angular/src/core/linker/component_factory.dart 86:51  ComponentFactory.create
...
```

As far as I know this parameter syntax has been legal since 1.24.0. Regardless, if I `pub build` and patch two such offending `Function` parameter types in the generated code (replacing, e.g., `void Function()` by just `Function`), then I can get component **tests to pass** with this combo:

- ✅ SDK **2.0.0-dev.17** (the latest)
- ✅ Angular 5.0.0-alpha+2
- ⚠️ Web-compiler: dart2js (+ **_manual patching_ of generated code**)
- ✅ Test platform: content-shell

### 3. Local (MacOS) using chrome (after patching generated code) ✅

**5.0.0-alpha+3 requires that we switch to using chrome** as a test platform, so that is the other variant in the setup, so I tried chrome under alpha+2 and tests pass; i.e., this combo works too:

- ✅ SDK 2.0.0-dev.17 (the latest)
- ✅ Angular 5.0.0-alpha+2
- ⚠️ Web-compiler: dart2js (+ manual patching of generated code)
- ✅ Test platform: **chrome**

### 4. Angular 5.0.0-alpha+3 using _dart2js_ ✅

Locally, component tests run successfully with this combo:

- ✅ SDK 2.0.0-dev.17
- ✅ Angular 5.0.0-**alpha+3**
- ✅ Web-compiler: dart2js (w/o patching of generated code)
- ✅ Test platform: chrome

### 5. Angular 5.0.0-alpha+3 using _dartdevc_ FAILS ❌

Use of this combo

- SDK 2.0.0-dev.17
- Angular 5.0.0*alpha+3
- ❌ Web-compiler: **dartdevc**
- Test platform: chrome

results in the errors shown in the opening comment.

The 5.0.0-alpha+3 changelog states that chrome + dartdevc should be used during development. I'll temporarily switch back to using dart2js to see if I can get all of our component tests passing again. Maybe the changelog entry should be adjusted to say that dart2js can (should?) still be used during development?

cc @kwalrath

**Comment by chalin** [comment]
Extra notes:

- I believe that the error reported over `void Function(E) eventHandler0<E>(void Function() handler)` is because the **content-shell** doesn't understand the new 1.24 Function syntax.

- While I can successfully run component tests under SDK dev.17 and Angular alpha+3, it only works if I launch `pub serve` separately. Letting angular_test launch `pub serve` results in:
  ```nocode
  Bad state: Should only get this output once.
  package:angular_test/src/bin/runner.dart 78  run.<fn>.<fn>
  ```

**Comment by alorenzen** [comment]
Do you still run into this issue when using `pub run build_runner test`, or only with `pub run angular_test`?

**Comment by matanlurey** [comment]
We figured this out elsewhere, externally DDC has stopped supporting (any) `dart:mirrors`.

@vsmenon: This is related to pkg/pageloader. How should we proceed?

**Comment by chalin** [comment]
FYI, when I run tests from the command line now, it reports "**No tests ran**":

```console
pub run build_runner test --fail-on-severe --delete-conflicting-outputs --output=build; echo $?
[INFO] Entrypoint: Generating build script...
[INFO] Entrypoint: Generating build script completed, took 326ms

[INFO] BuildDefinition: Initializing inputs
[INFO] BuildDefinition: Building new asset graph...
[INFO] BuildDefinition: Building new asset graph completed, took 595ms

[INFO] BuildDefinition: Checking for unexpected pre-existing outputs....
[INFO] BuildDefinition: Checking for unexpected pre-existing outputs. completed, took 1ms

[INFO] Build: Running build...
[INFO] Heartbeat: ... still running (5.1s so far)
...
[INFO] Build: Running build completed, took 20.8s

[INFO] Build: Caching finalized dependency graph...
[INFO] Build: Caching finalized dependency graph completed, took 124ms

[INFO] CreateOutputDir: Creating merged output dir `build`...
[INFO] CreateOutputDir: Creating merged output dir `build` completed, took 1.7s

[INFO] Build: Succeeded after 22.8s with 2331 outputs

Running tests...

No tests ran.
0
```

If I build and then run manually, I get the same error.

**Comment by kevmoo** [comment]
Which version of pkg/test are you getting? Latest?

**Comment by matanlurey** [comment]
@kevmoo This is due to DDC no longer supporting mirrors, not pkg/test.

**Comment by matanlurey** [comment]
I'm proposing in https://github.com/dart-lang/angular/issues/826 we remove PL support.

**Comment by matanlurey** [comment]
Closing as infeasible. See #826, we are removing support for pageloader at this time.

Once PL3 is available externally, we can re-add support.
